### PR TITLE
[[ Bug 23016 ]] Update emscripten custom font handling

### PR DESCRIPTION
**Pull request by livecodeian**
This patch updates the handling of custom fonts in the emscripten engine. Now font files added to the standalone bundle will be scanned to provide the list of custom fonts available (via the 'fontnames' property)

Fixes https://quality.livecode.com/show_bug.cgi?id=23016

### CODE REVIEW

File: engine/src/customfont.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+void MCCustomFontListAddFont(MCCustomFont *p_font);
+bool MCCustomFontListLookupFontByName(MCStringRef p_name, MCCustomFont* &r_font);
+bool MCCustomFontListLookupFontByFamilyAndStyle(MCStringRef p_family, bool p_bold, bool p_italic, MCCustomFont* &r_font);
+bool MCCustomFontListLookupFont(MCStringRef p_name, bool p_bold, bool p_italic, MCCustomFont* &r_font);
+
+bool MCCustomFontListGetNames(MCStringRef& r_names);
+MCCustomFontStyle MCCustomFontListGetStylesForName(MCStringRef p_name);
+
+#endif // __MC_CUSTOMFONT_H__
```
Comment (runrevmark) [inline engine/src/customfont.h]: Missing newline at end of file

File: engine/src/freetype-font.cpp [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+		r_family = t_family.Take();
+		r_style = t_style;
+	}
+	
+	if (t_font_face != nil)
+		FT_Done_Face(t_font_face);
+	
+	return t_success;
+}
+
```
Comment (runrevmark) [inline engine/src/freetype-font.cpp]: Extraneous newline at end of file

File: engine/src/freetype-font.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+You should have received a copy of the GNU General Public License
+along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
+
+#ifndef __MC_FREETYPE_FONT_H__
+#define __MC_FREETYPE_FONT_H__
+
+bool MCFreeTypeGetFontInfoFromData(FT_Library p_library, MCDataRef p_data, MCStringRef &r_name, MCStringRef &r_family, MCCustomFontStyle &r_style);
+
+#endif //__MC_FREETYPE_FONT_H__
```
Comment (runrevmark) [inline engine/src/freetype-font.h]: Missing newline at end of file

File: engine/src/em-font.cpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+		MCStringCreateWithCString(p_text, &t_string);
+		MCAutoStringRefAsUTF8String t_utf8_string;
+		/* UNCHECKED */ t_utf8_string . Lock(*t_string);
+		
+		/* UNCHECKED */ MCSkiaTypefaceMeasureText(t_font->typeface, t_font->size, *t_utf8_string, t_utf8_string.Size(), false, t_length);
+	}
+	
+	return t_length;
+}
+
```
Comment (runrevmark) [inline engine/src/em-font.cpp]: Extraneous newline at end of of file

File: engine/src/em-font.cpp [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+bool MCEmscriptenListFontFamilies(MCListRef& r_names)
+{
+	MCAutoListRef t_list;
+	if (!MCListCreateMutable('\n', &t_list))
+		return false;
+	
+	MCAutoStringRef t_custom_font_names;
+	if (!MCCustomFontListGetNames(&t_custom_font_names) ||
```
Comment (runrevmark) [inline engine/src/em-font.cpp:121]: Either do:
```
if (!... ||
        !...)
    return false;
```
or
```
if (... ||
    ...)
{
...   
}
```
i.e. if no braces make sure the condition wraps one level deeper than the body.

File: engine/src/em-font.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+bool MCEmscriptenListFontFamilies(MCListRef& r_names);
+bool MCEmscriptenListFontsForFamily(MCStringRef p_family, uint32_t p_fsize, MCListRef& r_styles);
+
+void* MCEmscriptenFontCreate(MCStringRef p_name, uint32_t p_size, bool p_bold, bool p_italic);
+void MCEmscriptenFontDestroy(void *p_font);
+void MCEmscriptenFontGetMetrics(void *p_font, float &r_ascent, float &r_descent, float &r_leading, float &r_xheight);
+float MCEmscriptenFontMeasureText(void *p_font, const char *p_text, uint32_t p_text_length, bool p_is_unicode);
+
+#endif // __MC_EM_FONT_H__
```
Comment (runrevmark) [inline engine/src/em-font.h]: No newline at end of file

File: engine/src/em-fontlist.cpp [hunk 7]
```diff
@@ -252,47 +253,3 @@ MCFontlist::getfontstructinfo(MCNameRef & r_name,
 	return false;
 }
```
Comment (runrevmark) [inline engine/src/em-fontlist.cpp]: Extraneous newline at end of file

**Review by runrevmark — CHANGES_REQUESTED** [review]
There's a typo in the release note commit message - `fir` rather than `for`.

File: engine/src/customfont.h (hunk 1 above)
Comment (livecodeian) [reply engine/src/customfont.h]: Fixed

File: engine/src/freetype-font.cpp (hunk 2 above)
Comment (livecodeian) [reply engine/src/freetype-font.cpp]: Fixed

File: engine/src/freetype-font.h (hunk 3 above)
Comment (livecodeian) [reply engine/src/freetype-font.h]: Fixed

File: engine/src/em-font.cpp (hunk 4 above)
Comment (livecodeian) [reply engine/src/em-font.cpp]: Fixed

File: engine/src/em-font.cpp (hunk 5 above)
Comment (livecodeian) [reply engine/src/em-font.cpp:121]: Fixed

File: engine/src/em-font.h (hunk 6 above)
Comment (livecodeian) [reply engine/src/em-font.h]: Fixed

File: engine/src/em-fontlist.cpp (hunk 7 above)
Comment (livecodeian) [reply engine/src/em-fontlist.cpp]: Fixed

**Comment by livecodeian** [comment]
@runrevmark I've resolved the extra / missing newlines and indenting and fixed the typo in the commit message

**Review by runrevmark — APPROVED** [review]

**Comment by runrevmark** [comment]
@livecode-vulcan review ok aaeb46e

**Comment by livecode-vulcan** [comment]
💙 review by @runrevmark ok aaeb46e808ea4c41b341051e4ff0c6abc9cd26e8

**Comment by livecode-vulcan** [comment]
😎 test success aaeb46e808ea4c41b341051e4ff0c6abc9cd26e8

* try-community-armv6-android-sdk26_ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv6-android-sdk26_ndk16r15/builds/1046)
* try-community-armv7-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-armv7-android-ndk16r15/builds/454)
* try-community-arm64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-arm64-android-ndk16r15/builds/453)
* try-community-x86-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86-android-ndk16r15/builds/452)
* try-community-x86_64-android-ndk16r15: [success](https://vulcan.livecode.com/builders/try-community-x86_64-android-ndk16r15/builds/453)
* try-community-js-emscripten-sdk1.35: [success](https://vulcan.livecode.com/builders/try-community-js-emscripten-sdk1.35/builds/2571)
* try-community-universal-ios-iphoneos14.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.5/builds/48)
* try-community-universal-ios-iphoneos14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphoneos14.4/builds/100)
* try-community-universal-ios-iphonesimulator14.5: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.5/builds/48)
* try-community-universal-ios-iphonesimulator14.4: [success](https://vulcan.livecode.com/builders/try-community-universal-ios-iphonesimulator14.4/builds/99)
* try-community-universal-mac-macosx10.9: [success](https://vulcan.livecode.com/builders/try-community-universal-mac-macosx10.9/builds/2597)
* try-community-x86-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86-linux-debian8/builds/2587)
* try-community-x86_64-linux-debian8: [success](https://vulcan.livecode.com/builders/try-community-x86_64-linux-debian8/builds/2585)
* try-community-x86-win32: [success](https://vulcan.livecode.com/builders/try-community-x86-win32/builds/2622)
* try-community-x86_64-win32: [success](https://vulcan.livecode.com/builders/try-community-x86_64-win32/builds/2558)

**Pull request merged**